sherpa/astro/io/pyfits_backend.py

```diff
@@ -780,32 +819,38 @@ def get_rmf_data(arg, make_copy=False):
                 rowdata.append(rdata)
                 start = end
 
-        data['matrix'] = numpy.concatenate(rowdata)
+        matrix = numpy.concatenate(rowdata)
 
-    data['matrix'] = data['matrix'].astype(SherpaFloat)
+    data['matrix'] = matrix.astype(SherpaFloat)
 
     # Flatten f_chan and n_chan vectors into 1D arrays as crates does
-    # according to group
+    # according to group. This is not always needed, but annoying to
+    # check for so we always do it.
     #
-    if data['f_chan'].ndim > 1 and data['n_chan'].ndim > 1:
-        f_chan = []
-        n_chan = []
-        for grp, fch, nch, in zip(data['n_grp'], data['f_chan'],
-                                  data['n_chan']):
-            for i in range(grp):
-                f_chan.append(fch[i])
-                n_chan.append(nch[i])
-
-        # This automatically filters out rows where N_GRP is 0.
-        #
-        data['f_chan'] = numpy.asarray(f_chan, SherpaUInt)
-        data['n_chan'] = numpy.asarray(n_chan, SherpaUInt)
-    else:
-        if len(data['n_grp']) == len(data['f_chan']):
-            # filter out groups with zeroes.
-            good = data['n_grp'] > 0
-            data['f_chan'] = data['f_chan'][good]
-            data['n_chan'] = data['n_chan'][good]
+    xf_chan = []
+    xn_chan = []
+    for grp, fch, nch, in zip(n_grp, f_chan, n_chan):
+        if numpy.isscalar(fch):
+            # The assumption is that grp==1 here
+            xf_chan.append(fch)
+            xn_chan.append(nch)
+        else:
+            # The arrays may contain extra elements (which
+            # should be 0).
+            #
+            xf_chan.extend(fch[:grp])
+            xn_chan.extend(nch[:grp])
+
+    data['f_chan'] = numpy.asarray(xf_chan, SherpaUInt)
+    data['n_chan'] = numpy.asarray(xn_chan, SherpaUInt)
+
+    # Not all fields are "flattened" by this routine. In
+    # particular we need to keep knowledge of these rows
+    # with 0 groups.
```

Comment:
I thought we removed those with zero groups with `good = data['n_grp'] > 0` above?

Reply:
So, the documentation and code behavior is a little sus here (he says, trying to sound cool), and my additions probably haven't made it clearer. I'll have to look more closely on Monday.

Reply:
I have updated the documentation to point out that we are not using `n_grp` here (which has had the 0's removed from it) but `data['n_grp']` (which retains the zeros). Does it make more sense now?